File: Mini-Parser/MP.go
package main

import (
	"fmt"
	"strconv"
)

type NestedInteger struct {
	value   int
	NestedL *NestedInteger
}

func deserialize(s string) *NestedInteger {
	str := []byte(s)
	if str[0] != '[' {
		val, _ := strconv.Atoi(string(str))
		a := NestedInteger{value: val}
		return &a
	} else {
		ss := str[1 : len(str)-1]
		//fmt.Print("this string: ")
		//fmt.Println(string(ss))
		var inTemp []byte
		var breakInd int

	lo:
		for i, v := range ss {
			fmt.Println(string(v))
			switch string(v) {
			case "0", "1", "2", "3", "4", "5", "6", "7", "8", "9":
				inTemp = append(inTemp, v)
			case "[":
				breakInd = i
				break lo
			}
		}

		val, _ := strconv.Atoi(string(inTemp))
		nextss := ss[breakInd:]

		a := NestedInteger{value: val, NestedL: deserialize(string(nextss))}
		return &a
	}
}

func main() {
	s1 := "324"
	s2 := "[123,[456,[789]]]"

	a := *deserialize(s1)
	b := *deserialize(s2)

	fmt.Println(a)
	fmt.Println(b)
}

File: Mini-Parser/MP.swift
class NestedInteger {
    var value:Int?
    var nestedL:NestedInteger?
}

func deserialize(_ s:String) -> NestedInteger {
    let a = NestedInteger()
    var str = Array(s.characters)

    if str[0] != "[" {
        let val = Int(s)!
        a.value = val
        return a
    }else{
        var ss = str[1..<str.endIndex-1]
        var breakInd = 1
        var inTemp:[Character] = []

        lo:for v in ss {
            switch v {
            case "0", "1", "2", "3", "4", "5", "6", "7", "8", "9":
                inTemp.append(v)
            case "[":
                breakInd = ss.index(of:v)!
                break lo
            default:
                continue
            }
        }

        let val = Int(String(inTemp))
        let nextss = String(ss[breakInd..<ss.endIndex])
        //print(nextss)

        a.value = val
        a.nestedL = deserialize(nextss)

        return a
    }
}

var s1 = "324"
var s2 = "[123,[456,[789]]]"


/**
print(deserialize(s1).value)
print(deserialize(s1).nestedL)
print(deserialize(s2).value)
print(deserialize(s2).nestedL!.value)
*/
